src/za/ac/sun/cs/green/service/simplify/ConstantPropogation.java

```diff
@@ -0,0 +1,129 @@
+package za.ac.sun.cs.green.service.simplify;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.SortedSet;
+import java.util.Stack;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.logging.Level;
+
+import za.ac.sun.cs.green.Instance;
+import za.ac.sun.cs.green.Green;
+import za.ac.sun.cs.green.expr.Expression;
+import za.ac.sun.cs.green.service.BasicService;
+import za.ac.sun.cs.green.util.Reporter;
+import za.ac.sun.cs.green.expr.Constant;
+import za.ac.sun.cs.green.expr.IntConstant;
+import za.ac.sun.cs.green.expr.IntVariable;
+import za.ac.sun.cs.green.expr.Operation;
+import za.ac.sun.cs.green.expr.Variable;
+import za.ac.sun.cs.green.expr.Visitor;
+import za.ac.sun.cs.green.expr.VisitorException;
+
+public class ConstantPropogation extends BasicService {
+
+	private int invocations = 0;
+
+	public ConstantPropogation(Green solver) {
+		super(solver);
+	}
+
+	@Override
+	public Set<Instance> processRequest(Instance instance) {
+		@SuppressWarnings("unchecked")
+		Set<Instance> result = (Set<Instance>) instance.getData(getClass());
+		if (result == null) {
+			final Map<Variable, Variable> map = new HashMap<Variable, Variable>();
+			final Expression e = canonize(instance.getFullExpression(), map);
+			final Instance i = new Instance(getSolver(), instance.getSource(), null, e);
+			result = Collections.singleton(i);
+			instance.setData(getClass(), result);
+		}
+		return result;
+	}
+
+	@Override
+	public void report(Reporter reporter) {
+		reporter.report(getClass().getSimpleName(), "invocations = " + invocations);
+	}
+
+	public Expression canonize(Expression expression, Map<Variable, Variable> map) {
+		try {
+			log.log(Level.FINEST, "Before Canonization: " + expression);
+			invocations++;
+			ConstantPropVisitor constantPropVisitor = new ConstantPropVisitor();
+			expression.accept(constantPropVisitor);
+			expression = constantPropVisitor.getExpression();
+			log.log(Level.FINEST, "After Canonization: " + expression);
+			return expression;
+		} catch (VisitorException x) {
+			log.log(Level.SEVERE, "encountered an exception -- this should not be happening!", x);
+		}
+		return null;
+	}
+
+	private static class ConstantPropVisitor extends Visitor {
+
+		private Stack<Expression> stack;
+		private HashMap<String, Integer> map;
+
+		public ConstantPropVisitor() {
+			stack = new Stack<Expression>();
```

Comment:
From the Java documentation for [`Stack`](https://docs.oracle.com/javase/8/docs/api/java/util/Stack.html):
> A more complete and consistent set of LIFO stack operations is provided by the[ `Deque` ](https://docs.oracle.com/javase/8/docs/api/java/util/Deque.html)interface and its implementations, which should be used in preference to this class.